zha/zigbee/cluster_handlers/__init__.py

```diff
@@ -624,6 +631,46 @@ async def write_attributes_safe(
                     f"Failed to write attribute {name}={value}: {record.status}",
                 )
 
+    async def _discover_attributes_all(
+        self,
+    ) -> list[DiscoverAttributesResponseRecord] | None:
+        discovery_complete = zigpy.types.Bool.false
+        start_attribute_id = 0
+        attribute_info = []
+        cluster = self.cluster
+        while discovery_complete != zigpy.types.Bool.true:
+            rsp = await cluster.discover_attributes(
+                start_attribute_id=start_attribute_id, max_attribute_ids=0xFF
+            )
+            assert rsp, "Must have a response to discover request"
+
+            if rsp.command.id == GeneralCommand.Default_Response:
+                self.debug(
+                    "Ignoring attribute discovery due to unexpected default response"
+                )
+                return None
+
+            attribute_info.extend(rsp.attribute_info)
+            discovery_complete = rsp.discovery_complete
+            start_attribute_id = (
+                max((info.attrid for info in rsp.attribute_info), default=0) + 1
+            )
+        return attribute_info
+
+    async def discover_unsupported_attributes(self):
+        """Discover the list of unsupported attributes from the device."""
+        attribute_info = await self._discover_attributes_all()
```

Comment:
Also i'm not fully sure what type of errors devices might respond with here. We may want to gracefully handle some things. For example, the GeneralCommand.Default_Response response is sent by some quirks here which seem utterly wrong.